### DISCUSSION TITLE
Legacy .net framework web application file access

### DISCUSSION THREAD

**Discussion by oysterbay73** [FantasticFiasco/serilog-sinks-http#241] [General]
Hi - Does anybody have experience of using this successfully with a legacy .net framework website. I cannot get it to work properly, it logs a few events and then (with self log on) I see:
`Serilog.Sinks.Http.Private.Network.HttpLogShipper: System.IO.IOException: The process cannot access the file '<myfilename>.bookmark' because it is being used by another process`.

The in memory sink works fine (as it uses no buffer file), but I have the same issue with both the file by time and file size sinks. I have tried using with/without the  bufferFileShared option. With it switched on, I also see that the last logged event before other events get blocked is sometimes logged many times (same event, same timestamp) to the http target (it is only present once in the buffer file). I've tried using an absolute path outside of the site in case some other process was causing the issue (and just to eliminate for testing granted everyone access), but that made no difference. The site is using framework 4.7.2 but I also tried downgrading it to no avail. I get the same issue using both IIS express and IIS, although it seems to take a little longer to go wrong in IIS express. 

It works absolutely fine in a Core (.NET 6.0) web app, and also using a .net framework console app, although the latter wouldn't be multi threaded which is, I assume, where the problem lies.

**Comment by FantasticFiasco** [comment]
It's not a problem I am aware of, but one I would like to resolve.

Would you be able to produce a GitHub repository with a sample application, and detailed instructions on how to run the application on IIS or IIS Express. Preferably some kind of script also that I can run to trigger the behavior.

Thanks!

**Reply by oysterbay73** [reply]
Thank you, I've created a repo here. 
https://github.com/oysterbay73/httpsink-demo

As detailed in the readme, there are 3 projects that demo the sink working, and the 1 framework web project where it does not. There is no valid http endpoint, I assume that you have 1 that you can use, but if not the problem is still evident from self log without.

If you toggle back and fore between the 'Contact' and 'Home' links when running then you should soon see the failure logged to debug output.

**Reply by FantasticFiasco** [reply]
The problem might relate to the differences in implementation when it comes to comparing `WebNetFw` with the other three applications. With `WebNetFw` you are re-instantiating the logger many times, while in the three other examples you only instantiate the logger once.

If we where to move the method `GetLogger` from `HomeController` to `Global.asax.cs`, and update the code to look like this, would you still see the error?

```csharp
protected void Application_Start()
{
    AreaRegistration.RegisterAllAreas();
    FilterConfig.RegisterGlobalFilters(GlobalFilters.Filters);
    RouteConfig.RegisterRoutes(RouteTable.Routes);
    BundleConfig.RegisterBundles(BundleTable.Bundles);

    Log.Logger = GetLogger();
}

private static Serilog.Core.Logger GetLogger()
{
    Serilog.Debugging.SelfLog.Enable(msg => Debug.WriteLine(msg));
    return new LoggerConfiguration()
        .MinimumLevel.Debug()
        .WriteTo.DurableHttpUsingFileSizeRolledBuffers(
            requestUri: "https://www.mylogs.com",
            httpClient: null,
            bufferBaseFileName: HostingEnvironment.MapPath("/buffer"),
            bufferFileShared: true) // required for multi threaded app - ensure app pool has full control
        .CreateLogger();
}
```

**Reply by oysterbay73** [reply]
Yes thanks very much, that works. I had only included in the controller for demo purposes, in my original project the `Microsoft.Extensions.Logging.ILogger` was being injected, so I guess I need to look at that, as it must also have instantiated each time.  Changes are pushed  up to my repo.

**Reply by oysterbay73** [reply]
Hmm, on further testing not quite there. It works fine in IIS express but when running in IIS it goes wrong after an IIS reset or app pool recycle. I think I had seen this before but not picked up on it properly due to the other error.

Start the site - works fine
Recycle the app pool - All appears well but if you look in the log file, the 1st record logged after the recycle has an invalid character in it at the start. When I have the site pointing to a valid http endpoint the batch sending of logs gets stuck here. I viewed the file in notepad++ with language set to json, this highlights the offending line.

I haven't been able to recreate the same issue when running the core web app under iis, it seems to happen on every recycle with the framework web app.

**Reply by FantasticFiasco** [reply]
Good, we're getting closer to the issue!

Do you think it might be the same issue as #174? In any way, would you be able to update your repo with detailed instructions on how to trigger the behavior in IIS. Assume I know nothing about IIS, how would I go about to use your repo to trigger the behavior in IIS?

One other thing. Next week is a vacation week for me, I won't be able to work on this issue until week 6. Thanks for understanding.

**Reply by oysterbay73** [reply]
OK thanks, and enjoy your break, skiing I'm going to guess from your location! I've added details for setting up IIS, the app pool, and binding to the site in the readme. The part I'm not 100% sure on is what options come enabled on IIS by default when you enable it, and it may vary based on Windows version and edition.

**Reply by oysterbay73** [reply]
I just googled IIS a bit more, I'm not sure if home edition windows comes with enough IIS components.

**Reply by FantasticFiasco** [reply]
No worries, AWS or Azure will provide the server with IIS, just get me the instructions and I will figure out the rest.

**Reply by FantasticFiasco** [reply]
Regarding the invalid character, it has been fixed in [v8.0.0-beta.4](https://github.com/FantasticFiasco/serilog-sinks-http/releases/tag/v8.0.0-beta.4). The pre-releases are of no lesser quality than the officially released v7.2.0, but we are in the process of stabilizing the API, thus the pre-release versions.

**Comment by oysterbay73** [comment]
That's great news,  thank you for looking into it
